### PR TITLE
Upgrade enhancements

### PR DESCRIPTION
**Pull request by detiber**
Supercedes https://github.com/openshift/openshift-ansible/pull/816

### CODE REVIEW

**Comment by dgoodwin** [comment]
This passed for me. Using this branch for a test build for QE tonight.

File: playbooks/adhoc/upgrades/upgrade.yml [hunk 1]
```diff
@@ -1,35 +1,66 @@
...
+  gather_facts: no
   tasks:
-  # Checking the global deployment type rather than host facts, this is about
-  # what the user is requesting.
-    - fail: msg="Deployment type enterprise not supported for upgrade"
+    # Checking the global deployment type rather than host facts, this is about
+    # what the user is requesting.
+    - fail:
+        msg: "Deployment type enterprise not supported for upgrade"
       when: deployment_type == "enterprise"
```
Comment (dgoodwin) [inline playbooks/adhoc/upgrades/upgrade.yml]: I sense this should be skipped if this is a 3.0 to 3.0.2 upgrade. I will get a version check on this.

File: playbooks/adhoc/upgrades/upgrade.yml [hunk 2]
```diff
@@ -43,62 +74,65 @@
...
+      msg: Atomic OpenShift 3.1 packages not found
+    when: deployment_type in ['openshift-enterprise', 'atomic-openshift'] and g_aos_versions.curr_version | version_compare('3.0.2.900','<') and (g_aos_versions.avail_version is none or g_aos_versions.avail_version | version_compare('3.0.2.900','<'))
 
-- name: Verify upgrade can proceed
-  hosts: oo_first_master
-  tasks:
-  # Checking the global deployment type rather than host facts, this is about
-  # what the user is requesting.
-  - fail: msg="Deployment type 'enterprise' must be updated to 'openshift-enterprise' for upgrade to proceed"
-    when: deployment_type == "enterprise" and (_new_version.stdout | version_compare('1.0.7', '>=') or _new_version.stdout | version_compare('3.1', '>='))
```
Comment (dgoodwin) [inline playbooks/adhoc/upgrades/upgrade.yml]: Ah here's the original for the above, we lost a little bit in translation. I will fix.

File: playbooks/adhoc/upgrades/upgrade.yml (hunk 1 above)
Comment (detiber) [inline playbooks/adhoc/upgrades/upgrade.yml]: Hmm, we're going to have bigger issues if we want to support both 3.0.z upgrades as well as 3.0.z -> 3.1 upgrades with the same playbook...

File: playbooks/adhoc/upgrades/upgrade.yml (hunk 2 above)
Comment (detiber) [inline playbooks/adhoc/upgrades/upgrade.yml]: We will most likely need multiple playbooks to handle different upgrades.

File: playbooks/adhoc/upgrades/upgrade.yml [hunk 3]
```diff
@@ -106,16 +140,139 @@
...
-      openshift_upgrade_config: from_version=3.0 to_version=3.1 role=master
-    - name: Restart master services
-      service: name="{{ openshift.common.service_type}}-master" state=restarted
+
+    - name: Ensure python-yaml present for config upgrade
+      yum:
+        pkg: python-yaml
+        state: installed
+
+    - debug: var=hostvars[inventory_hostname].openshift.common.config_base
```
Comment (dgoodwin) [inline playbooks/adhoc/upgrades/upgrade.yml]: I'm going to remove this as well.

File: playbooks/adhoc/upgrades/upgrade.yml [hunk 4]
```diff
@@ -106,16 +140,139 @@
...
+
+    - name: Upgrade master configuration
+      openshift_upgrade_config:
+        from_version: '3.0'
+        to_version: '3.1'
+        role: master
+        config_base: "{{ hostvars[inventory_hostname].openshift.common.config_base }}"
```
Comment (dgoodwin) [inline playbooks/adhoc/upgrades/upgrade.yml]: Probably neglects 3.0.x to 3.0.2, should I make this version conditional as well? (only if new version >= 3.1)

File: playbooks/adhoc/upgrades/upgrade.yml (hunk 3 above)
Comment (detiber) [inline playbooks/adhoc/upgrades/upgrade.yml]: Already done

**Comment by detiber** [comment]
@brenton This branch should now be up to date with the latest upgrade changes that @abutcher @dgoodwin and myself have been working on.

**Comment by brenton** [comment]
Very cool.  I'm running some basic sanity tests now.

**Comment by detiber** [comment]
@brenton once this has been merged, I'm planning on working up a new change that will break this playbook up and make different upgrades more distinct and clear (3.0 minor upgrades vs 3.0 -> 3.1 upgrade, etc)

**Pull request merged**